### DISCUSSION TITLE
Randomized Results when using the SSIM metric

### DISCUSSION THREAD

**Discussion by Ord-D** [ImageMagick/ImageMagick#3802] [Help]
Hello,

I'm trying to use SSIM-values in a paper, but I'm getting problem with the ImageMagick-results from the command: " magick compare -metric SSIM exampleA.png exampleB.png null"
The resulting values arent accurate, though they are close to the real value, and furthermore they are also slightly randomized each time:

PS C:\Users\little white> magick compare -metric SSIM 2_real_A.png 2_fake_B.png null; echo " "
0.80881
PS C:\Users\little white> magick compare -metric SSIM 2_real_A.png 2_fake_B.png null; echo " "
0.808734
PS C:\Users\little white> magick compare -metric SSIM 2_real_A.png 2_fake_B.png null; echo " "
0.802256
PS C:\Users\little white> magick compare -metric SSIM 2_real_A.png 2_fake_B.png null; echo " "
0.806921

Even one and the same picture end up weird:

PS C:\Users\little white> magick compare -metric SSIM 2_real_A.png 2_real_A.png null; echo " "
1.02665
PS C:\Users\little white> magick compare -metric SSIM 2_real_A.png 2_real_A.png null; echo " "
1.03364

I tried the SSIM-calculation on another tool repeatedly and got these values consistently:
0.856 for A and B
1.000 for A and A

Though other metrics seem to function as intended:

PS C:\Users\little white> magick compare -metric PSNR 2_real_A.png 2_fake_B.png null; echo " "
23.7538
PS C:\Users\little white> magick compare -metric PSNR 2_real_A.png 2_fake_B.png null; echo " "
23.7538

I'm using the Powershell of a Win7-32bit Computer, though I also tried it on my Win10-64bit Laptop.
The ImageMagick-Version I downloaded is ImageMagick-7.0.11-Q16 for Win.

Is there are problem on my side? Or is there some bug in ImageMagick?

**Comment by snibgo** [comment]
Perhaps this is a recent bug in IM. For versions 7.0.8-64 and 7.0.10-45, the results are constant.

**Reply by fmw42** [reply]
I also get slight variations from run to run using IM 7.0.11.14 Q16 Mac OSX Sierra. This does not happen with my own script version that does it from scratch.

**Reply by Ord-D** [reply]
I tried one of these older versions and now I'm getting consistent results. So i guess there is a bug in the newer versions?
Anyway, thanks for your help! Now I should be able to continue with the rest of my paper.

**Comment by dlemstra** [comment]
Thanks for reporting this. I just pushed a patch to fix this in the next release of ImageMagick.
